WordPress/gutenberg
# e2e test with selenium + browserstack for legacy browsers like IE.

* Closes #16509 ?

## Description

* Experimental implementation of e2e testing with selenium + browserstack for legacy browsers like Internet Explorer 11. 
* I tested the package setup by writing e2e test code for #21276. I know it's more suitable for a separate PR. But I want to be sure if it's a worthwhile attempt.

## How has this been tested?

1. Turn on WordPress dev environment by `npx wp-env start`.
2. Build plugin by `npm run dev` or `npm run build`.
3. Go to `packages/e2e-selenium` folder. 
4. Copy `.env.template` file and rename it to `.env`. Fill the content of `.env`. You can get the name and key by signing up browserstack. If you have any problem signing up or don't want to sign up for this, I'll share my username and key. Contact sainthkh at WordPress Slack.
5. Run `../../node_modules/.bin/jest --config=jest.config.js`.

After a few minutes, you can see the message like below:

![image](https://user-images.githubusercontent.com/8130013/78952843-d1a2f680-7b11-11ea-85a4-803dc301fd84.png)

Then, you can watch the result by visiting [the link](https://automate.browserstack.com/builds/011fae724222641159b4fa92e5b2dec6b9954965/sessions/582a258b280006e37cfa2a993c6a5140c77c6222?auth_token=07725e55937cc588ebcada35674fe72300b9f54e5cbf8f718eeb3e257ae983bc). (Login is not required.)

## Screenshots 

Check above.

## Types of changes

New testing feature.

## Things to discuss.

### 1\. How to show the result link.

Currently, I used `console.log()` to log it. But it's against the Gutenberg convention. `process.stdout.write()` is overwritten by the jest result message. It cannot be used. I also used `process.stdout.on('data')` and `process.on('exit')`, but they didn't work.
 
My last option is to use [`child_process.fork()`](https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options) and IPC. But I hope there is simpler answer than this. If there's no other way than this, I need to use it.

### 2\. Moving the selenium+browserstack configuartion to packages/scripts?

I'm not sure this setup is useful for WordPress developers.

### 3\. How to organize tests.

I created a folder, `ie`, under `test` folder. And create a test file with an issue number. It's because it takes really long to finish a test (10-20s a test or even longer). By breaking down a test file per issue, we can test them separately. 

Downside: It's hard to see what's going on from the name of the file. But I guess issue page contains more info than the file name. So, it don't think it's important.

### 4\. Setting up BrowserStack and Travis CI.

To make this work with CI, we need an official BrowserStack account for Gutenberg team and [set up environment variables.](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings)

I created `CI_BROWSERSTACK_USER`, `CI_BROWSERSTACK_KEY` for them.

## Checklist:
- [N/A] My code follows the accessibility standards. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.
